File: freeq-ios/freeq/Models/AvatarCache.swift
import SwiftUI

/// Fetches and caches Bluesky avatar URLs from the public API.
@MainActor
class AvatarCache: ObservableObject {
    static let shared = AvatarCache()

    @Published private var cache: [String: URL] = [:]  // nick -> avatar URL
    private var dids: [String: String] = [:]  // nick -> DID (from message account tags etc.)
    private var pending: Set<String> = []
    private var failed: Set<String> = []  // Don't retry failed lookups

    /// Get cached avatar URL for a nick. Returns nil if not yet fetched.
    func avatarURL(for nick: String) -> URL? {
        cache[nick]
    }

    /// The DID last seen for a nick (from a message account tag, account-notify,
    /// or WHOIS). Lets identity UIs resolve a profile for senders whose DID
    /// arrived on a message rather than in the member roster (e.g. custom-domain
    /// handles like chadfowler.com, or federated senders).
    func did(for nick: String) -> String? {
        dids[nick.lowercased()]
    }

    /// Request avatar fetch for a nick. Resolution requires a verified
    /// `did` — see `fetchAvatar` for why we never resolve from the nick.
    /// A no-DID call is a no-op: we simply wait for the DID to arrive
    /// (account-tag on a message, account-notify, or WHOIS) and resolve then.
    func prefetch(_ nick: String, did: String? = nil) {
        let key = nick.lowercased()
        // Skip guest nicks - they're not Bluesky accounts.
        guard !key.hasPrefix("guest"), !key.hasPrefix("web") else { return }
        if cache[key] != nil || pending.contains(key) { return }
        // Identity on freeq is the DID the server bound at SASL — never the
        // freely-settable nick. Without a verified DID there is nothing we
        // can safely resolve, and `did:key` users (guests, AI beings) have
        // no Bluesky profile at all.
        guard let did = did, !did.isEmpty, !did.hasPrefix("did:key:") else { return }
        dids[key] = did
        if failed.contains(key) { return }
        pending.insert(key)

        Task {
            await fetchAvatar(nick: nick, key: key, did: did)
        }
    }

    /// Prefetch avatars for a list of nicks.
    func prefetchAll(_ nicks: [String]) {
        for nick in nicks {
            prefetch(nick)
        }
    }

    private func fetchAvatar(nick: String, key: String, did: String? = nil) async {
        // Resolve ONLY by the server-verified DID. We must never derive a
        // Bluesky identity from the nick — neither the bare nick as a handle
        // nor a guessed "<nick>.bsky.social". Nicks are freely chosen, so any
        // such guess shows a STRANGER's photo and handle for whoever happens
        // to match (e.g. the AI being "olive" pulling up the unrelated real
        // account olive.bsky.social). That is impersonation. `prefetch`
        // already guarantees a non-empty, non-did:key DID here.
        guard let did = did, !did.isEmpty else {
            failed.insert(key)
            pending.remove(key)
            return
        }
        if let url = await resolveAvatar(handle: did) {
            cache[key] = url
        } else {
            failed.insert(key)
        }
        pending.remove(key)
    }

    private func resolveAvatar(handle: String) async -> URL? {
        let urlString = "https://public.api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=\(handle.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? handle)"
        guard let url = URL(string: urlString) else { return nil }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            if let avatarStr = json?["avatar"] as? String, let avatarURL = URL(string: avatarStr) {
                return avatarURL
            }
        } catch { }
        return nil
    }
}

/// SwiftUI view that displays a user's avatar (cached Bluesky profile pic or initial).
struct UserAvatar: View {
    let nick: String
    let size: CGFloat
    @StateObject private var cache = AvatarCache.shared

    var body: some View {
        Group {
            if let url = cache.avatarURL(for: nick.lowercased()) {
                AsyncImage(url: url) { image in
                    image.resizable().scaledToFill()
                } placeholder: {
                    initialCircle
                }
                .frame(width: size, height: size)
                .clipShape(Circle())
            } else {
                initialCircle
                    .onAppear { cache.prefetch(nick) }
            }
        }
    }

    private var initialCircle: some View {
        ZStack {
            // Subtle top-lit gradient of the member's signature color — matches
            // the colored name in the transcript, with a little depth.
            Circle()
                .fill(
                    LinearGradient(
                        colors: [nickColor, nickColor.opacity(0.72)],
                        startPoint: .top, endPoint: .bottom)
                )
                .frame(width: size, height: size)
                .overlay(
                    Circle().strokeBorder(.white.opacity(0.14), lineWidth: 1)
                )
            Text(String(nick.prefix(1)).uppercased())
                .font(.system(size: size * 0.42, weight: .semibold, design: .rounded))
                .foregroundColor(Color(hex: "04121a").opacity(0.88))
        }
    }

    /// One consistent color system for a member — the same hash the transcript
    /// uses for their name (Theme.nickColor), so avatar and name always agree.
    private var nickColor: Color {
        Theme.nickColor(for: nick)
    }
}

File: freeq-android/freeq/src/main/java/com/freeq/model/AvatarCache.kt
package com.freeq.model

import kotlinx.coroutines.*
import org.json.JSONObject
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.ConcurrentHashMap

data class BlueskyProfile(
    val handle: String,
    val displayName: String?,
    val description: String?,
    val avatar: String?,
    val followersCount: Int?,
    val followsCount: Int?,
    val postsCount: Int?
)

object AvatarCache {
    private val cache = ConcurrentHashMap<String, String>()  // nick -> avatar URL
    private val profileCache = ConcurrentHashMap<String, BlueskyProfile>()
    private val pending = ConcurrentHashMap.newKeySet<String>()
    private val failed = ConcurrentHashMap.newKeySet<String>()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    fun avatarUrl(nick: String): String? = cache[nick.lowercase()]

    fun profile(nick: String): BlueskyProfile? = profileCache[nick.lowercase()]

    suspend fun fetchProfileIfNeeded(nick: String, did: String?): BlueskyProfile? {
        val key = nick.lowercase()
        profileCache[key]?.let { return it }
        // DID-only: without a verified, non-did:key DID there is no Bluesky
        // profile to show. See fetchAvatar for why we never resolve from a nick.
        if (did.isNullOrEmpty() || did.startsWith("did:key:")) return null
        pending.add(key)
        fetchAvatar(key, did)
        return profileCache[key]
    }

    fun prefetch(nick: String, did: String? = null) {
        val key = nick.lowercase()
        // Skip guest nicks - they're not Bluesky accounts.
        if (key.startsWith("guest") || key.startsWith("web")) return
        if (cache.containsKey(key) || pending.contains(key)) return
        // Identity on freeq is the DID the server bound at SASL — never the
        // freely-settable nick. Without a verified DID there is nothing we can
        // safely resolve, and did:key users (guests, AI beings) have no Bluesky
        // profile. A no-DID call is a no-op (and NOT marked failed) so a later
        // call carrying the account-tag DID is never pre-empted.
        if (did.isNullOrEmpty() || did.startsWith("did:key:")) return
        if (failed.contains(key)) return
        pending.add(key)
        scope.launch { fetchAvatar(key, did) }
    }

    fun prefetchAll(nicks: List<String>) {
        nicks.forEach { prefetch(it) }
    }

    private suspend fun fetchAvatar(key: String, did: String) {
        // Resolve ONLY by the server-verified DID. We must never derive a
        // Bluesky identity from the nick — neither the bare nick as a handle
        // nor a guessed "<nick>.bsky.social". Nicks are freely chosen, so any
        // such guess shows a STRANGER's photo and handle for whoever happens to
        // match (e.g. the AI being "olive" pulling up the unrelated real
        // account olive.bsky.social). That is impersonation.
        val result = resolveProfile(did)
        if (result != null) {
            profileCache[key] = result
            result.avatar?.let { cache[key] = it }
        } else {
            failed.add(key)
        }
        pending.remove(key)
    }

    // Test seam: unit tests override this to assert we only ever resolve by a
    // verified DID (never a nick or a guessed "<nick>.bsky.social"). Defaults
    // to the live Bluesky API call.
    internal var resolveProfile: (String) -> BlueskyProfile? = ::resolveProfileNetwork

    /** Clear all caches and restore the live resolver. Test-only. */
    internal fun resetForTest() {
        cache.clear(); profileCache.clear(); pending.clear(); failed.clear()
        resolveProfile = ::resolveProfileNetwork
    }

    private fun resolveProfileNetwork(handle: String): BlueskyProfile? {
        return try {
            val encoded = URLEncoder.encode(handle, "UTF-8")
            val url = URL("https://public.api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=$encoded")
            val conn = url.openConnection().apply {
                connectTimeout = 5000
                readTimeout = 5000
            }
            val text = conn.getInputStream().bufferedReader().readText()
            val json = JSONObject(text)
            BlueskyProfile(
                handle = json.optString("handle", handle),
                displayName = json.opt("displayName") as? String,
                description = json.opt("description") as? String,
                avatar = json.opt("avatar") as? String,
                followersCount = if (json.has("followersCount")) json.optInt("followersCount") else null,
                followsCount = if (json.has("followsCount")) json.optInt("followsCount") else null,
                postsCount = if (json.has("postsCount")) json.optInt("postsCount") else null
            )
        } catch (_: Exception) {
            null
        }
    }
}
